container.te

```diff
@@ -672,6 +672,13 @@ kernel_read_network_state(container_t)
 kernel_dontaudit_write_proc_files(container_t)
 kernel_get_sysvipc_info(container_t)
 
+gen_require(`
+	type unconfined_t;
+')
+allow container_t unconfined_t:sem { associate getattr unix_read };
```

Comment:
I am not crazy about these ones, and these would only happen if you run the container under an unconfined domain. What command did you run to generate these AVCs?

Reply:
This actually looks like a bug when you use --ipc host, you are not supposed to be running as container_t, it should be running as spc_t.  Upstream Docker might have broken some functionality here.

Reply:
So the issue is with how Kubernetes (and potentially other CRI implementations such as cri-containerd and quite possibly cri-o too) implement Pods.

They implement it by running a "sandbox" container (running a "pause" binary) that sets up network and IPC...

Then the other containers in the Pod just point at that container for their setup.

In order to reproduce something like this, these commands can be used:

```
    $ ipcmk -M 1M -S 10 -Q
    Shared memory id: 98304
    Message queue id: 65536
    Semaphore id: 131072
    $ docker run --name hostipc_pod --detach --ipc host busybox sleep 3600
    $ docker run --ipc container:hostipc_pod busybox ipcs
```

So, indeed, container "hostipc_pod" *is* running as spc_t, and running `ipcs` there (from busybox) works. But the second container is running as container_t, so `ipcs` fails there.

Perhaps the answer is for Docker to somewhat figure out that the second container is sharing some of the resources (network, IPC, etc.) from an spc_t container and so it should set spc_t there too?

Reply:
BTW, this works fine:

```
$ docker run --ipc container:hostipc_pod --security-opt label:type:spc_t busybox ipcs

------ Message Queues --------
key        msqid      owner      perms      used-bytes   messages    
0x02563ed9 65536      5192       644        0            0           

------ Shared Memory Segments --------
key        shmid      owner      perms      bytes      nattch     status      
0xbed1a300 98304      5192       644        1048576    0                       

------ Semaphore Arrays --------
key        semid      owner      perms      nsems     
0xf87f2222 131072     5192      644        10        
```

And I think this could be made into a fix, since Kubernetes knows that the Pod requested HostIPC so it could request to run in SELinux type "spc_t". I'm not sure whether relying on a type named "spc_t" is OK though...

Of course, another option is just to run all containers in the pod with `--ipc host` instead of making them point to the sandbox for their network and IPC... (In fact, that was my first approach in kubernetes/kubernetes#58174.) But I'm not sure whether there would be unexpected/undesirable side effects from doing that. Opinions?

Cheers,
Filipe

Reply:
docker run --ipc container:hostipc_pod  busybox cat /proc/self/attr/current
Should be the same SELinux label as hostipc_pod.